### PR TITLE
feat(modp2p): Make IP Colocation configurable for pubsub parameters

### LINKED ISSUE #1842: modp2p: add Allowlist configuration for IPColocation PubSub parameter
There is [TODO](https://github.com/celestiaorg/celestia-node/blob/main/nodebuilder/p2p/pubsub.go#L136-L141) that explains what has to be done.

### PR DESCRIPTION
**Pull request by derrandz**
## Overview

Closes #1842 

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

### CODE REVIEW

File: nodebuilder/p2p/pubsub.go [hunk 1]
```diff
@@ -109,12 +110,23 @@ func topicScoreParams(network Network) map[string]*pubsub.TopicScoreParams {
...
+	var ipColocFactWl []*net.IPNet = nil
+	if len(cfg.IPColocationFactorWhitelist) > 0 {
+		for _, strIp := range cfg.IPColocationFactorWhitelist {
+			_, ipNet, err := net.ParseCIDR(strIp)
+			if err != nil {
+				panic(err)
+			}
+			ipColocFactWl = append(ipColocFactWl, ipNet)
+		}
+	}
```
Comment (Wondertan) [inline nodebuilder/p2p/pubsub.go:128]: Some simplifications:
* No need to check for an empty slice before iterating, as the range is noop in such cases
* No nee to explicitly set nil when using var
```suggestion
	var ipColocFactWl []*net.IPNet
	for _, strIp := range cfg.IPColocationFactorWhitelist {
		_, ipNet, err := net.ParseCIDR(strIp)
		if err != nil {
			panic(err)
		}
		ipColocFactWl = append(ipColocFactWl, ipNet)
	}
```

File: nodebuilder/p2p/pubsub.go [hunk 2]
```diff
@@ -109,12 +110,23 @@ func topicScoreParams(network Network) map[string]*pubsub.TopicScoreParams {
...
 	for _, b := range bootstrappers {
 		bootstrapperSet[b.ID] = struct{}{}
 	}
 
+	var ipColocFactWl []*net.IPNet = nil
+	if len(cfg.IPColocationFactorWhitelist) > 0 {
+		for _, strIp := range cfg.IPColocationFactorWhitelist {
+			_, ipNet, err := net.ParseCIDR(strIp)
+			if err != nil {
+				panic(err)
```
Comment (Wondertan) [inline nodebuilder/p2p/pubsub.go:124]: We should not panic but return a nice error to the user, as it's clearly a misconfiguration by the user

**Review by Wondertan — CHANGES_REQUESTED** [review]

File: nodebuilder/p2p/pubsub.go [hunk 3]
```diff
@@ -109,12 +115,21 @@ func topicScoreParams(network Network) map[string]*pubsub.TopicScoreParams {
...
 	}
 
+	var ipColocFactWl []*net.IPNet
+	for _, strIp := range cfg.IPColocationFactorWhitelist {
+		_, ipNet, err := net.ParseCIDR(strIp)
+		if err != nil {
+			return nil, fmt.Errorf("error while parsing string ip: %+v", err)
+		}
+		ipColocFactWl = append(ipColocFactWl, ipNet)
+	}
```
Comment (Wondertan) [inline nodebuilder/p2p/pubsub.go:131]: Can we preallocate the number net.IPNet?

**Review by Wondertan — APPROVED** [review]

File: nodebuilder/p2p/pubsub.go (hunk 3 above)
Comment (Wondertan) [reply nodebuilder/p2p/pubsub.go:131]: Ah, yes, linter also complains about it

**Comment by Wondertan** [comment]
Please also make the name feat(modp2p) and set the labels

File: nodebuilder/p2p/config.go [hunk 4]
```diff
@@ -35,6 +35,9 @@ type Config struct {
 	// ConnManager is a configuration tuple for ConnectionManager.
 	ConnManager               connManagerConfig
 	RoutingTableRefreshPeriod time.Duration
+
+	// Allowlist for IPColocation PubSub parameter, a list of string IPs
+	IPColocationFactorWhitelist []string
```
Comment (Wondertan) [inline nodebuilder/p2p/config.go:40]:
```suggestion
	// Allowlist for IPColocation PubSub parameter, a list of string CIDRs
	IPColocationWhitelist []string
```

File: nodebuilder/p2p/pubsub.go [hunk 5]
```diff
@@ -109,12 +115,21 @@ func topicScoreParams(network Network) map[string]*pubsub.TopicScoreParams {
...
 	bootstrapperSet := map[peer.ID]struct{}{}
 	for _, b := range bootstrappers {
 		bootstrapperSet[b.ID] = struct{}{}
 	}
 
+	var ipColocFactWl []*net.IPNet //nolint
+	for _, strIP := range cfg.IPColocationFactorWhitelist {
+		_, ipNet, err := net.ParseCIDR(strIP)
+		if err != nil {
+			return nil, fmt.Errorf("error while parsing string ip: %+v", err)
```
Comment (Wondertan) [inline nodebuilder/p2p/pubsub.go:128]:
```suggestion
			return nil, fmt.Errorf("error while parsing whitelist collocation CIDR string: %w", err)
```

File: nodebuilder/p2p/pubsub.go [hunk 6]
```diff
@@ -109,12 +115,21 @@ func topicScoreParams(network Network) map[string]*pubsub.TopicScoreParams {
...
 }
 
-func peerScoreParams(isBootstrapper bool, bootstrappers Bootstrappers) *pubsub.PeerScoreParams {
+func peerScoreParams(isBootstrapper bool, bootstrappers Bootstrappers, cfg Config) (*pubsub.PeerScoreParams, error) {
 	bootstrapperSet := map[peer.ID]struct{}{}
 	for _, b := range bootstrappers {
 		bootstrapperSet[b.ID] = struct{}{}
 	}
 
+	var ipColocFactWl []*net.IPNet //nolint
```
Comment (Wondertan) [inline nodebuilder/p2p/pubsub.go:124]: Why not preallocate?

**Review by Wondertan — CHANGES_REQUESTED** [review]

File: nodebuilder/p2p/pubsub.go [hunk 7]
```diff
@@ -109,12 +115,21 @@ func topicScoreParams(network Network) map[string]*pubsub.TopicScoreParams {
...
+	for _, strIP := range cfg.IPColocationFactorWhitelist {
```
Comment (Wondertan) [inline nodebuilder/p2p/pubsub.go:125]: Applying rename suggestion over GH does not change it where it's used so this broken now

**Review by Wondertan — CHANGES_REQUESTED** [review]

File: nodebuilder/p2p/pubsub.go [hunk 8]
```diff
@@ -109,12 +115,21 @@ func topicScoreParams(network Network) map[string]*pubsub.TopicScoreParams {
...
 }
 
-func peerScoreParams(isBootstrapper bool, bootstrappers Bootstrappers) *pubsub.PeerScoreParams {
+func peerScoreParams(isBootstrapper bool, bootstrappers Bootstrappers, cfg Config) (*pubsub.PeerScoreParams, error) {
 	bootstrapperSet := map[peer.ID]struct{}{}
 	for _, b := range bootstrappers {
 		bootstrapperSet[b.ID] = struct{}{}
 	}
 
+	ipColocFactWl := make([]*net.IPNet, 0)
```
Comment (Wondertan) [inline nodebuilder/p2p/pubsub.go:124]: It's not a finished preallocation
```suggestion
	ipColocFactWl := make([]*net.IPNet, 0, len(cfg.IPColocationWhitelist))
```

**Review by Wondertan — CHANGES_REQUESTED** [review]

**Comment by Wondertan** [comment]
> Please also make the name feat(modp2p) and set the labels

**Comment by derrandz** [comment]
Sorry didn't see that comment, thanks! @Wondertan

**Comment by Wondertan** [comment]
@derrandz, all our PRs have to have a `kind:` label, as there is CI rule check for this, which you can see it's broken

**Review by Wondertan — APPROVED** [review]

**Review by renaynay — APPROVED** [review]

**Pull request merged**